pkg/formats/sniffer.go

```diff
@@ -25,72 +34,105 @@ func (fs *Sniffer) SniffReader(f io.ReadSeeker) (Format, error) {
 	fileScanner := bufio.NewScanner(f)
 	fileScanner.Split(bufio.ScanLines)
 
-	formatType := ""
-	formatEncoding := ""
-	formatVersion := ""
+	var format formatDetails
 
 	for fileScanner.Scan() {
-		if strings.Contains(fileScanner.Text(), `"bomFormat"`) && strings.Contains(fileScanner.Text(), `"CycloneDX"`) {
-			formatType = "application/vnd.cyclonedx"
-			formatEncoding = JSON
+		format = fs.sniff(fileScanner.Bytes())
+		if format.Version != "" && format.Type != "" && format.Encoding != "" {
+			break
 		}
+	}
 
-		if strings.Contains(fileScanner.Text(), `"specVersion"`) {
-			parts := strings.Split(fileScanner.Text(), ":")
-			if len(parts) == 2 {
-				ver := strings.TrimPrefix(strings.TrimSuffix(strings.TrimSuffix(strings.TrimSpace(parts[1]), ","), "\""), "\"")
-				if ver != "" {
-					formatVersion = ver
-					formatEncoding = JSON
-				}
-			}
+	fmt.Fprintf(
+		os.Stderr, "format: %s version: %s encoding: %s\n",
+		format.Type, format.Version, format.Encoding,
+	)
+
+	for _, f := range List {
+		if string(f) == fmt.Sprintf("%s+%s;version=%s", format.Type, format.Encoding, format.Version) {
+			return f, nil
 		}
+	}
 
-		if strings.Contains(fileScanner.Text(), "SPDXVersion:") {
-			formatType = "text/spdx"
-			formatEncoding = "text"
+	// TODO(puerco): Implement a light parser in case the string hacks don't work
+	return "", fmt.Errorf("unknown SBOM format")
+}
 
-			for _, ver := range []string{"2.2", "2.3"} {
-				if strings.Contains(fileScanner.Text(), fmt.Sprintf("SPDX-%s", ver)) {
-					formatVersion = ver
-					break
-				}
-			}
-			break
+func (fs *Sniffer) sniff(data []byte) formatDetails {
+	for _, sig := range sniffFormats {
+		if format := sig.sniff(data); format.Type != "" {
+			return format
 		}
+	}
+
+	return formatDetails{}
+}
```

Comment:
This is the function I was referring to. The previous algorithm captures less than one read the SBOM format. THe format gets detected in almost all cases by reading the first few lines.

In this new function, the format sniffer will reads the whole SBOM entirely for each failed format before doing now finally pass to find the version.

The example I noted was for the k8s SBOM. It will read the whole 18 MB file, detect it is not CycloneDX and then it needs to read it again from the beginning to find and return the SPDX version. And we're not taking into account XML or tag value yet.

Reply:
Sorry for the missunderstanding
This function is called by parent on each file fragment.
The function sniff them recives the fragment in the bytes , and runs each sniffer only on the fragment them returns on the first found format so as the parent caller,  does come it actually going over the entire file?

Reply:
Ok, i looked in to it, i think i now understand what your reffering to.

i had a bug i did not save the partial state of the format  (i was assuming the fragment is assumed to include all the details (type,name and version) for each format type.
I commits a fix for this specific bug (not all your comments resolved in this commit).
Does this resolve your concern about the IO reads?

I for some reason assumed the file fragment includes all the format details in one block (eg 512 byte block or something), i now tested it and found that it goes line by line and sometime the version is found in a separate line than the encoding.
I should never assume things sorry..